### PR TITLE
fix docker

### PR DESCRIPTION
**Pull request by shaoyaoqian**
I am trying to build the Docker image with the Dockerfile in the docker/ directory. There are some problems, and I am still working on them.

It works now.

### CODE REVIEW

**Comment by shaoyaoqian** [comment]
```log
[+] Building 371.2s (14/16)                                                                                   docker:default
 => [internal] load .dockerignore                                                                                       0.0s
 => => transferring context: 94B                                                                                        0.0s
 => [internal] load build definition from Dockerfile                                                                    0.0s
 => => transferring dockerfile: 1.58kB                                                                                  0.0s
 => [internal] load metadata for ghcr.io/fenics/dolfinx/lab:v0.8.0                                                      0.0s
 => [ 1/12] FROM ghcr.io/fenics/dolfinx/lab:v0.8.0                                                                      0.0s
 => [internal] load build context                                                                                       0.0s
 => => transferring context: 36B                                                                                        0.0s
 => CACHED [ 2/12] WORKDIR /tmp/                                                                                        0.0s
 => CACHED [ 3/12] RUN apt-get update && apt-get install -y libgl1-mesa-glx libxrender1 xvfb curl                       0.0s
 => [ 4/12] RUN curl -sL https://deb.nodesource.com/setup_18.x -o nodesource_setup.sh &&     bash nodesource_setup.sh  21.8s
 => [ 5/12] RUN python3 -m pip install -U setuptools pip pkgconfig                                                    146.4s
 => [ 6/12] RUN echo linux/amd64                                                                                        0.4s 
 => [ 7/12] RUN if [ "linux/amd64" = "linux/arm64" ]; then python3 -m pip install "https://github.com/finsberg/vtk-aar  0.5s 
 => [ 8/12] RUN if [ "linux/amd64" = "linux/amd64" ]; then python3 -m pip install vtk; fi                               2.1s 
 => [ 9/12] ADD pyproject.toml /tmp/pyproject.toml                                                                      0.2s 
 => ERROR [10/12] RUN python3 -m pip install --no-cache-dir --no-binary=h5py -v .                                     181.3s 
------                                                                                                                       
 > [10/12] RUN python3 -m pip install --no-cache-dir --no-binary=h5py -v .:                                                  
0.835 Using pip 24.1.2 from /usr/lib/python3/dist-packages/pip (python 3.10)                                                 
0.900 Processing /tmp                                                                                                        
0.904   Installing build dependencies: started                                                                               
0.904   Running command pip subprocess to install build dependencies
1.284   Using pip 24.1.2 from /usr/lib/python3/dist-packages/pip (python 3.10)
34.11   Collecting setuptools>=64.4.0
34.11     Obtaining dependency information for setuptools>=64.4.0 from https://files.pythonhosted.org/packages/ef/15/88e46eb9387e905704b69849618e699dc2f54407d8953cc4ec4b8b46528d/setuptools-70.3.0-py3-none-any.whl.metadata
34.45     Downloading setuptools-70.3.0-py3-none-any.whl.metadata (5.8 kB)
35.66   Collecting wheel
35.66     Obtaining dependency information for wheel from https://files.pythonhosted.org/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl.metadata
35.77     Downloading wheel-0.43.0-py3-none-any.whl.metadata (2.2 kB)
36.06   Collecting pip>=22.3
36.06     Obtaining dependency information for pip>=22.3 from https://files.pythonhosted.org/packages/e7/54/0c1c068542cee73d8863336e974fc881e608d0170f3af15d0c0f28644531/pip-24.1.2-py3-none-any.whl.metadata
36.06     Using cached pip-24.1.2-py3-none-any.whl.metadata (3.6 kB)
36.17   Downloading setuptools-70.3.0-py3-none-any.whl (931 kB)
36.91      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 931.1/931.1 kB 1.3 MB/s eta 0:00:00
37.03   Downloading wheel-0.43.0-py3-none-any.whl (65 kB)
37.06      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.8/65.8 kB 2.1 MB/s eta 0:00:00
37.06   Using cached pip-24.1.2-py3-none-any.whl (1.8 MB)
37.41   Installing collected packages: wheel, setuptools, pip
37.46     Creating /tmp/pip-build-env-2_i8ecfi/overlay/bin
37.46     changing mode of /tmp/pip-build-env-2_i8ecfi/overlay/bin/wheel to 755
38.98     changing mode of /tmp/pip-build-env-2_i8ecfi/overlay/bin/pip to 755
38.99     changing mode of /tmp/pip-build-env-2_i8ecfi/overlay/bin/pip3 to 755
38.99     changing mode of /tmp/pip-build-env-2_i8ecfi/overlay/bin/pip3.10 to 755
39.01   Successfully installed pip-24.1.2 setuptools-70.3.0 wheel-0.43.0
39.01   WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.
39.19   Installing build dependencies: finished with status 'done'
39.19   Getting requirements to build wheel: started
39.20   Running command Getting requirements to build wheel
39.43   running egg_info
39.43   creating DOLFINx_Tutorial.egg-info
39.43   writing DOLFINx_Tutorial.egg-info/PKG-INFO
39.43   writing dependency_links to DOLFINx_Tutorial.egg-info/dependency_links.txt
39.43   writing requirements to DOLFINx_Tutorial.egg-info/requires.txt
39.43   writing top-level names to DOLFINx_Tutorial.egg-info/top_level.txt
39.43   writing manifest file 'DOLFINx_Tutorial.egg-info/SOURCES.txt'
39.43   reading manifest file 'DOLFINx_Tutorial.egg-info/SOURCES.txt'
39.43   writing manifest file 'DOLFINx_Tutorial.egg-info/SOURCES.txt'
39.47   Getting requirements to build wheel: finished with status 'done'
39.47   Preparing metadata (pyproject.toml): started
39.47   Running command Preparing metadata (pyproject.toml)
39.69   running dist_info
39.70   creating /tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info
39.70   writing /tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/PKG-INFO
39.70   writing dependency_links to /tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/dependency_links.txt
39.70   writing requirements to /tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/requires.txt
39.70   writing top-level names to /tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/top_level.txt
39.70   writing manifest file '/tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/SOURCES.txt'
39.70   reading manifest file '/tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/SOURCES.txt'
39.70   writing manifest file '/tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial.egg-info/SOURCES.txt'
39.70   creating '/tmp/pip-modern-metadata-du3u5obf/DOLFINx_Tutorial-0.8.0.dist-info'
39.74   Preparing metadata (pyproject.toml): finished with status 'done'
55.66 Collecting jupyter-book (from DOLFINx_Tutorial==0.8.0)
55.66   Obtaining dependency information for jupyter-book from https://files.pythonhosted.org/packages/c6/a4/644d35f1b038940d25f8b457ba1939506b8e0f22f4360a2ac2d4fc2250d0/jupyter_book-1.0.2-py3-none-any.whl.metadata
56.16   Downloading jupyter_book-1.0.2-py3-none-any.whl.metadata (5.8 kB)
56.88 Collecting meshio (from DOLFINx_Tutorial==0.8.0)
56.88   Obtaining dependency information for meshio from https://files.pythonhosted.org/packages/89/5f/39cbadc320cd78f4834b0a9f7a2fa3c980dca942bf193f315837eacb8870/meshio-5.3.5-py3-none-any.whl.metadata
57.03   Downloading meshio-5.3.5-py3-none-any.whl.metadata (11 kB)
57.35 Collecting h5py (from DOLFINx_Tutorial==0.8.0)
57.46   Downloading h5py-3.11.0.tar.gz (406 kB)
57.85      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 406.5/406.5 kB 1.1 MB/s eta 0:00:00
57.94   Running command pip subprocess to install build dependencies
57.94   Installing build dependencies: started
58.33   Using pip 24.1.2 from /usr/lib/python3/dist-packages/pip (python 3.10)
58.39   Ignoring oldest-supported-numpy: markers 'python_version == "3.8"' don't match your environment
118.5   WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ConnectTimeoutError(<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f25b6f4f730>, 'Connection to pypi.org timed out. (connect timeout=15)')': /simple/cython/
171.1   ERROR: Could not find a version that satisfies the requirement Cython<4,>=0.29.31 (from versions: none)
171.1   ERROR: No matching distribution found for Cython<4,>=0.29.31
171.2   error: subprocess-exited-with-error
171.2   
171.2   × pip subprocess to install build dependencies did not run successfully.
171.2   │ exit code: 1
171.2   ╰─> See above for output.
171.2   
171.2   note: This error originates from a subprocess, and is likely not a problem with pip.
171.2   full command: /usr/bin/python3 /usr/lib/python3/dist-packages/pip/__pip-runner__.py install --ignore-installed --no-user --prefix /tmp/pip-build-env-jqie969c/overlay --no-warn-script-location -v --no-binary h5py --only-binary :none: -i https://pypi.org/simple -- 'Cython >=0.29.31,<4' 'oldest-supported-numpy; python_version=='"'"'3.8'"'"'' 'numpy >=2.0.0rc1; python_version>='"'"'3.9'"'"'' pkgconfig 'setuptools >=61'
171.2   cwd: [inherit]
171.2   Installing build dependencies: finished with status 'error'
171.2 error: subprocess-exited-with-error
171.2 
171.2 × pip subprocess to install build dependencies did not run successfully.
171.2 │ exit code: 1
171.2 ╰─> See above for output.
171.2 
171.2 note: This error originates from a subprocess, and is likely not a problem with pip.
------
Dockerfile:34
--------------------
  32 |     
  33 |     ADD pyproject.toml /tmp/pyproject.toml
  34 | >>> RUN python3 -m pip install --no-cache-dir --no-binary=h5py -v .
  35 |     RUN python3 -m pip cache purge
  36 |     RUN jupyter lab build
--------------------
ERROR: failed to solve: process "/bin/sh -c python3 -m pip install --no-cache-dir --no-binary=h5py -v ." did not complete successfully: exit code: 1
```

File: README.md [hunk 1]
```diff
@@ -46,7 +46,7 @@ An image building DOLFINx, Basix, UFL and FFCx from source can be built using:
 
 ```bash
 cd docker
-docker build -f ./docker/Dockerfile -t local_lab_env .
+docker build -f ./Dockerfile -t local_lab_env .
```
Comment (jorgensd) [inline README.md:49]: This instructs the user to build the outermost docker file, which is not required. I would suggest reverting this.

File: docker/Dockerfile [hunk 2]
```diff
@@ -14,7 +14,7 @@ WORKDIR /tmp/
 RUN apt-get update && apt-get install -y libgl1-mesa-glx libxrender1 xvfb curl
 RUN curl -sL https://deb.nodesource.com/setup_18.x -o nodesource_setup.sh && \
     bash nodesource_setup.sh && \
-    apt install nodejs
+    apt -y install nodejs
```
Comment (jorgensd) [inline docker/Dockerfile:17]: Wonder how i missed this.

File: README.md (hunk 1 above)
Comment (shaoyaoqian) [reply README.md:49]: It is duplicated. One of the following works.
```
cd docker
docker build -f ./Dockerfile -t local_lab_env .
```
```
docker build -f ./docker/Dockerfile -t local_lab_env .
```

Comment (jorgensd) [reply README.md:49]: Right. I would then go for:
1. remove the cd
2. Use the original instruction


Reasoning behind this is that then the context of the Dockerfile is the root of the repo

File: README.md [hunk 3]
```diff
@@ -46,7 +46,7 @@ An image building DOLFINx, Basix, UFL and FFCx from source can be built using:
 
 ```bash
 cd docker
```
Comment (jorgensd) [inline README.md:48]:
```suggestion
```

**Pull request merged**